src/misc/rwqueue.cpp

```diff
@@ -124,6 +124,18 @@ bool RWQueue<T>::Enqueue(T&& item)
 	return is_running;
 }
 
+template <typename T>
+bool RWQueue<T>::NonblockingEnqueue(T&& item)
+{
+	std::lock_guard<std::mutex> lock(mutex);
+	if (!is_running || queue.size() >= capacity) {
+		return false;
+	}
+	queue.push_back(std::move(item));
+	has_items.notify_one();
+	return true;
+}
```

Comment:
Suggest moving the notification outside from the lock, something like:
```c++
	std::unique_lock<std::mutex> lock(mutex);
	if (!is_running || queue.size() >= capacity) {
		return false;
	}
	queue.push_back(std::move(item));

	lock.unlock();
	has_items.notify_one();
	return true;
```

Because if all notification are locked then it can lead to a deadlock. Notifications are thread safe (and the pattern in the other code is to explicitly unlock prior to notification).

Reply:
I don't think this will ever deadlock.  It's just slightly inefficient (the waiting thread could block on the mutex until it gets released one line later).  I'll change it.

Reply:
(yeah, maybe not critical; but the docs recommend notifying outside the lock too - see point 3. https://en.cppreference.com/w/cpp/thread/condition_variable)

Reply:
Fixed